core/src/main/scala/io/odin/loggers/AsyncLogger.scala

```diff
@@ -13,39 +13,44 @@ import scala.concurrent.duration._
   *
   * Use `AsyncLogger.withAsync` to instantiate it safely
   */
-case class AsyncLogger[F[_]](queue: ConcurrentQueue[F, LoggerMessage], timeWindow: FiniteDuration, inner: Logger[F])(
-    implicit F: Concurrent[F],
-    timer: Timer[F],
-    contextShift: ContextShift[F]
+case class AsyncLogger[F[_]](queue: Queue[F, LoggerMessage], timeWindow: FiniteDuration, inner: Logger[F])(
+    implicit F: Async[F]
 ) extends DefaultLogger[F](inner.minLevel) {
   def submit(msg: LoggerMessage): F[Unit] = {
     queue.tryOffer(msg).void
   }
 
   def drain: F[Unit] = {
-    queue
-      .drain(0, Int.MaxValue)
+    drainAll
       .flatMap { msgs =>
         inner.log(msgs.toList)
       }
-      .void
-      .handleErrorWith { _ =>
-        F.unit
-      }
+      .orElse(F.unit)
   }
 
   /**
     * Run internal loop of consuming events from the queue and push them down the chain
     */
-  def runF: F[Fiber[F, Unit]] = {
-    def recDrain: F[Unit] =
-      drain >> timer.sleep(timeWindow) >> contextShift.shift >> recDrain
-    F.start(recDrain).map { fiber =>
-      Fiber(fiber.join, drain >> fiber.cancel)
+  def runF: F[Fiber[F, Throwable, Unit]] = {
+    def drainOnce: F[Unit] = drain >> F.sleep(timeWindow) >> F.cede
+
+    F.start(drainOnce.foreverM[Unit]).map { fiber =>
```

Comment:
there is a `Supervisor` abstraction in CE3: 

https://github.com/typelevel/cats-effect/blob/series/3.x/std/shared/src/main/scala/cats/effect/std/Supervisor.scala

Would it fit better here?

Reply:
Also https://github.com/typelevel/cats-effect/blob/series/3.x/kernel/shared/src/main/scala/cats/effect/kernel/GenSpawn.scala

Reply:
I'm not sure to be honest. `runF` already is a part of the `AsyncLogger` lifecycle.

Btw, should `runF` even be public or part of the class? The event consumer loop should be started only once.

At this case, the definition of the `AsyncLogger` can be simplified:

```scala
case class AsyncLogger[F: Monad: Clock] private (...) extends DefaultLogger(...) {
  def submit(msg: LoggerMessage): F[Unit] = ...
  private def drain: F[Unit] = ...
}

object AsyncLogger {
  def withAsync[F[_]: Async](inner: Logger[F], timeWindow: FiniteDuration, maxBufferSize: Option[Int]): Resource[F, Logger[F]] = {
    val createQueue = ...
    
    def backgroundConsumer(logger: AsyncLogger[F]): Resource[F, Unit] = {
      def drainLoop: F[Unit] = F.andWait(logger.drain, timeWindow).foreverM[Unit]
    
      // cannot use F.background due to a custom cancellation logic
      Resource.make(F.start(drainLoop))(fiber => logger.drain >> fiber.cancel).void
    }

    for {
      queue  <- Resource.eval(createQueue)
      logger <- Resource.pure(AsyncLogger(queue, timeWindow, inner))
      _      <- backgroundConsumer
    } yield logger
  }  
}
```

Reply:
sounds good to me